lib/jekyll-remote-theme/dependency_manager.rb

```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+module Jekyll
+  module RemoteTheme
+    # Since evaluating a gemspec from third-party themes can lead to arbitrary code
+    # execution, this class safely reads a gemspec file at the root of given theme
+    # and handles any runtime dependencies declared within.
+    class DependencyManager
+      DEPENDENCY_MATCHER = %r!add_(?:runtime_)?dependency!
+      EXTRACT_DEPENDENCY_REGEX = %r!#{DEPENDENCY_MATCHER}(?:[(|\s])["'](.*?[^\\])["']!
+
+      # Returns an array of dependency gem-names.
+      attr_reader :theme_dependencies
+
+      def initialize(theme, whitelist)
+        @theme = theme
+        @whitelist = whitelist
+        @theme_dependencies = []
+      end
+
+      # Reads the gemspec at the given path, line-by-line, checking if the line
+      # contains a dependency declaration.
+      #
+      # Returns nothing, but stores the match in an array which can be accessed by
+      # calling +:theme_dependencies+
+      def extract_dependencies
+        return if @gemspec.nil? || !@theme_dependencies.empty?
+
+        File.read(@gemspec).each_line do |line|
+          next unless line =~ DEPENDENCY_MATCHER
+          line.match(EXTRACT_DEPENDENCY_REGEX)
+          @theme_dependencies << Regexp.last_match(1)
+        end
+      end
+
+      # Traverse the array of dependencies and +require+ the dependency if its
+      # whitelisted for current site.
+      #
+      # Returns nothing
+      def require_dependencies
+        theme_dependencies.each do |dependency|
+          next if dependency == "jekyll"
+          if @whitelist.include?(dependency)
```

Comment:
The idea of the plugin maintaining its own copy of the plugin whitelist and implementing its own logic feels very feels insecure and prone to logic getting out of sync (e.g., this doesn't respect the safe mode flag). Can we not use [`plugin_allowed`](https://github.com/jekyll/jekyll/blob/9ec9273ed9f04a666e6bc434baa2790321675c8f/lib/jekyll/plugin_manager.rb#L68)? 

Even better, could we set `theme.runtime_dependencies` and use [`require_theme_deps`](https://github.com/jekyll/jekyll/blob/9ec9273ed9f04a666e6bc434baa2790321675c8f/lib/jekyll/plugin_manager.rb#L35-L44)?

Reply:
> safe mode flag

I left it out since its always `safe: true` on GHP..
To use `plugin_allowed`, I'll need a copy of the `PluginManager` instance
I can't directly use `require_theme_deps` because there, each entry of `site.theme.runtime_dependencies` responds to a  `:name` method, whereas here individual entries are just regular strings..